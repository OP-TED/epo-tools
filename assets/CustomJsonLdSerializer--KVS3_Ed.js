import{S as e}from"./index-D8tqZlAQ.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
