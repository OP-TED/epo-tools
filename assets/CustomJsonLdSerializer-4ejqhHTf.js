import{S as e}from"./index-BzUeugQh.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
