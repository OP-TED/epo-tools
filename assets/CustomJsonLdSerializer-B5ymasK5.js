import{S as e}from"./index-C3U9xmz_.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
