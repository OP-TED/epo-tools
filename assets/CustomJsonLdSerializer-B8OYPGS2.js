import{S as e}from"./index-B8eoWoro.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
