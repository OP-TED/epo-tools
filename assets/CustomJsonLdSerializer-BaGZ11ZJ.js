import{S as e}from"./index-C_l3gxd7.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
