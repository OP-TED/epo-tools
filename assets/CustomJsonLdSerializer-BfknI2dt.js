import{S as e}from"./index-B-0oM0QG.js";class t extends e{constructor({...r}={}){super({encoding:"string",...r})}}export{t as default};
