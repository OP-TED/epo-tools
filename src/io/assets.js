import formats from '@rdfjs/formats'
import fs from 'fs'
import { glob } from 'glob'
import rdf from 'rdf-ext'

function getMimetype (path) {
  if (path.endsWith('.rdf')) {
    return 'application/rdf+xml'
  }
  return 'text/turtle'
}

function getGraph (path) {
  return rdf.namedNode(`file://${path}`)
}

async function applyGlob (globPattern) {
  const files = await glob(globPattern, {
    stat: true, nodir: true,
  })

  if (files.length === 0) {
    console.log('No files found')
    console.log({ workingDir: process.cwd(), globPattern })
  } else {
    console.log('loading', files.length, 'files')
  }
  return files
}

async function getRdfAssets ({ globPattern }, graphFactory = getGraph) {

  const turtleFiles = await applyGlob(globPattern)

  const assets = []
  for (const path of turtleFiles) {

    const fileStream = fs.createReadStream(path, 'utf-8')
    const dataset = rdf.dataset()
    try {
      const mimeType = getMimetype(path)
      await dataset.import(formats.parsers.import(mimeType, fileStream))
      for (const quad of [...dataset]) {
        quad.graph = graphFactory(path)
      }
    } catch (e) {
      // console.error(path, 'failed to parse')
      throw e
    }
    assets.push({ path, dataset })
  }
  return assets
}

export { getRdfAssets, applyGlob }
