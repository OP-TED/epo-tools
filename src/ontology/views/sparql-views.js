import { Parser } from 'sparqljs'

const sparqlQuery = `
PREFIX epo: <http://data.europa.eu/a4g/ontology#>
PREFIX cccev: <http://data.europa.eu/m8g/>
PREFIX dct: <http://purl.org/dc/terms/> 

SELECT DISTINCT
?Notice
?PublicationDate
?AwardedValue
?LotAwardedValueCurrencyURI
where {

    VALUES ?PublicationDate {
        "20230911"
    }

    ?Notice
            epo:hasPublicationDate ?PublicationDate;
            epo:refersToLot ?Lot .
            ?Lot a epo:Lot .
    ?LotAwardOutcome epo:describesLot ?Lot ;
                     epo:hasAwardedValue / epo:hasAmountValue ?AwardedValue;
                     epo:hasAwardedValue / epo:hasCurrency ?LotAwardedValueCurrencyURI.

} LIMIT 10
`

const parser = new Parser()
const parsedQuery = parser.parse(sparqlQuery)

// console.log(JSON.stringify(parsedQuery, null, 2))

// Function to extract named nodes
function extractNamedNodes (query) {
  const namedNodes = []
  const prefixes = query.prefixes

  query.where.forEach(element => {
    if (element.type === 'bgp' && Array.isArray(element.triples)) {
      element.triples.forEach(triple => {
        if (triple.predicate && triple.predicate.termType === 'NamedNode') {
          const predicateValue = triple.predicate.value
          const prefixMatch = Object.entries(prefixes).
            find(([prefix, uri]) => predicateValue.startsWith(uri))
          if (prefixMatch) {
            const [prefix, uri] = prefixMatch
            const curie = `${prefix}:${predicateValue.substring(uri.length)}`
            namedNodes.push(curie)
          } else {
            namedNodes.push(predicateValue) // No prefix, use full URI
          }
        }
      })
    }
  })

  return namedNodes
}

const namedNodes = extractNamedNodes(parsedQuery)
console.log(namedNodes)


