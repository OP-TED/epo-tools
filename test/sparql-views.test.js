import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { readFileSync } from 'fs'
import { describe, it } from 'mocha'
import { applyGlob } from '../src/io/assets.js'

import { termsFromQuery } from '../src/ontology/views/sparql-views.js'

expect.extend({ toMatchSnapshot })

const globPattern = `test/support/queries/**`
const assets = await applyGlob(globPattern)

describe('sparql-views', async () => {

  for (const path of assets) {
    const queryStr = readFileSync(path).toString()
    it(`asset:${path}`, async function () {
      const { terms } = termsFromQuery({ queryStr })
      expect(terms).toMatchSnapshot(this)
    })
  }

  it(`failed query`, async function () {
    const { terms, error } = termsFromQuery({ queryStr: 'not a query' })
    expect(error).toMatchSnapshot(this)
  })
})

