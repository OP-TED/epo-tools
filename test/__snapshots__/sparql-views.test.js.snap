// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sparql-views asset:test/support/queries/SMEs_by_country.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:announcesRole",
  "epo:Winner",
  "epo:playedBy",
  "epo:hasBusinessSize",
]
`;

exports[`sparql-views asset:test/support/queries/highest_value_of_contract.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:describesLot",
  "epo:hasAwardedValue",
  "epo:comprisesTenderAwardOutcome",
  "epo:TenderAwardOutcome",
  "org:Organization",
  "epo:hasLegalName",
  "epo:MonetaryValue",
  "epo:hasAmountValue",
  "epo:hasCurrency",
]
`;

exports[`sparql-views asset:test/support/queries/mets_packages.rq 1`] = `
Array [
  "cdm:procurement_public",
  "cdm:work_date_document",
  "owl:sameAs",
  "cmr:creationDate",
  "cmr:lastModificationDate",
  "cdm:expression_belongs_to_work",
  "cdm:expression_uses_language",
  "cdm:manifestation_manifests_expression",
  "cdm:manifestation_type",
  "cdm:item_belongs_to_manifestation",
]
`;

exports[`sparql-views asset:test/support/queries/notice.rq 1`] = `Array []`;

exports[`sparql-views asset:test/support/queries/notices_by_buyers_country.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:announcesRole",
  "epo:Buyer",
  "epo:playedBy",
  "epo:hasLegalName",
]
`;

exports[`sparql-views asset:test/support/queries/tenders_foresee_subcontracting.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:isSubmittedForLot",
  "epo:foreseesSubcontracting",
]
`;

exports[`sparql-views asset:test/support/queries/values_of_tenders.rq 1`] = `
Array [
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:describesLot",
]
`;

exports[`sparql-views asset:test/support/queries/winners.rq 1`] = `
Array [
  "epo:Notice",
  "epo:hasPublicationDate",
  "epo:refersToLot",
  "epo:describesLot",
  "epo:hasAwardedValue",
  "epo:comprisesTenderAwardOutcome",
  "epo:TenderAwardOutcome",
  "org:Organization",
  "epo:hasLegalName",
  "epo:MonetaryValue",
  "epo:hasAmountValue",
  "epo:hasCurrency",
]
`;

exports[`sparql-views failed query 1`] = `
[Error: Parse error on line 1:
not a query
^
Expecting 'EOF', 'BASE', 'IRIREF', 'PREFIX', 'PNAME_NS', 'CONSTRUCT', 'DESCRIBE', 'ASK', 'SELECT', '(', 'LOAD', 'CREATE', 'INSERTDATA', 'DELETEDATA', 'DELETEWHERE', 'DELETE', 'INSERT', 'WITH', 'a', '!', '^', 'PNAME_LN', 'CLEAR', 'DROP', 'ADD', 'MOVE', 'COPY', got 'NOT']
`;
